FileGopherServer.cpp

```diff
@@ -0,0 +1,260 @@
+    void *adr = mmap(NULL, s.st_size, PROT_READ, MAP_SHARED, fd, 0); // i get the adress
+    write(client_fd, adr, s.st_size);                                // i send the file from this adress directly
+}
+
+/**
+ * Function to get contents of selector on the server.
+ * The selector is contained in buffer.
+ * The contents are also placed in buffer and size of 
+ * contents is returned.
+ * @param buffer The selector.
+ * @param size The size of selector.
+ * @return The size of contents.
+ */
+int FileGopherServer::get_contents(char buffer[], int size)
+{
+    // Remove CRLF from selector.
+    if (buffer[size - 2] == CR && buffer[size - 1] == LF)
+    {
+        buffer[size - 2] = '\0';
+        size -= 2;
+    }
+    string route(buffer, buffer + size);
+    // Check if selector is gopher+.
+    if (buffer[0] == '\t')
+    {
+        if (routes.find(route) == routes.end())
+        {
+            strcpy(buffer, "\0");
+            return 0;
+        }
+        strcpy(buffer, routes[route].c_str());
+        return routes[route].size();
+    }
+    // Full path of selector.
+    string dir = "";
+    dir += root + route;
+    printf("client requested %s\n", dir.c_str());
+    // If the path is a file, send it.
+    if (is_regular_file(dir.c_str()))
+    {
+        // The path is a file, so send the file contents instead
+        strcpy(buffer, dir.c_str());
+        return -1;
+    }
+    // Read the contents of directory.
+    string files;
+    DIR *dp;
+    struct dirent *dirp;
+    if ((dp = opendir(dir.c_str())) == NULL)
+    {
+        cout << "Error(" << errno << ") opening " << dir << endl;
+        return 0;
+    }
+    while ((dirp = readdir(dp)) != NULL)
+    {
+        // Skip hidden files
+        // TODO: Add parameter to show hidden files
+        if (dirp->d_name[0] == '.')
+            continue;
+        string filename = dir + "/" + string(dirp->d_name);
+        files += create_line(
+            (dirp->d_type == DT_DIR) ? '1' : get_file_type(filename.c_str()),
+            dirp->d_name,
+            route + "/" + string(dirp->d_name),
+            socket_host,
+            socket_port);
+    }
+    closedir(dp);
+    strcpy(buffer, files.c_str());
+    return files.size();
+}
+
+/********* Public Functions **********/
+
+/**
+ * Function to create a gopher listing.
+ * @param type The type of listing.
+ * @param user_string The string visible to user.
+ * @param path The selector to the listing.
+ * @param host The host where the listing resides.
+ * @param port The port where the listing resides.
+ */
+string FileGopherServer::create_line(char type, string user_string,
+                                     string path, string host, int port)
+{
+    // Convert integer port to string.
+    char ch_port[6];
+    sprintf(ch_port, "%d", port);
+    return string(1, type) + user_string + TAB + path + TAB + host + TAB + string(ch_port) + CRLF;
```

Comment:
Cut this line so that it comes within 80 char/line limit

Reply:
Done

Reply:
Is it?

I can't see the change